libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/GroupAllocator.kt

```diff
@@ -0,0 +1,51 @@
+package net.corda.messaging.mediator
+
+import net.corda.messaging.api.mediator.config.EventMediatorConfig
+import net.corda.messaging.api.records.Record
+import kotlin.math.ceil
+
+/**
+ * Helper class to use in the mediator to divide polled records into groups for processing.
+ */
+class GroupAllocator {
+
+    /**
+     * Allocate events into groups based on their keys, a configured minimum group size and thread count.
+     * This allows for more efficient multi-threaded processing.
+     * The threshold record count to establish a new group is [config.minGroupSize].
+     * If the number of groups exceeds the number of threads then the group count is set to the number of [config.threads]
+     * Records of the same key are always placed into the same group regardless of group size and count.
+     * @param events Events to allocate to groups
+     * @param config Mediator config
+     * @return Records allocated to groups.
+     */
+    fun <K : Any, S : Any, E : Any> allocateGroups(
+        events: List<Record<K, E>>,
+        config: EventMediatorConfig<K, S, E>
+    ): List<Map<K, List<Record<K, E>>>> {
+        val groups = setUpGroups(config, events)
+        val buckets = events.groupBy { it.key }
+        val bucketsSized = buckets.keys.sortedByDescending { buckets[it]?.size }
+        for (i in 0 until buckets.size) {
+            val group = groups.minBy { it.values.flatten().size }
+            val key = bucketsSized[i]
+            group[key] = buckets[key]!!
+        }
+
+        return groups.filter { it.values.isNotEmpty() }
+    }
```

Comment:
```suggestion
    fun <K : Any, S : Any, E : Any> allocateGroups(
        events: List<Record<K, E>>,
        config: EventMediatorConfig<K, S, E>
    ): List<Map<K, List<Record<K, E>>>> {
        val groups = setUpGroups(config, events)
        val buckets = events
            .groupBy { it.key }.toList()
            .sortedByDescending { it.second.size }
        
        buckets.forEach { (key, records) ->
            val leastFilledGroup = groups.minByOrNull { it.values.flatten().size }
            leastFilledGroup?.put(key, records)
        }
        
        return groups.filter { it.values.isNotEmpty() }
    }
```
Removes the iterator-based for loop and the dreaded `!!`